### PR TITLE
[WebNN EP] Fix bug in Softmax

### PR DESCRIPTION
**Pull request by zesongw**
### Description

For now, WebNN Softmax only support 2D (or implicitly coerce to 2D) inputs and the last axis.

### Motivation and Context

Fallback some cases to pass the CI.

### CODE REVIEW

**Comment by zesongw** [comment]
@fdwr, @guschmue, PTAL, thanks!

**Review by guschmue — APPROVED** [review]

File: onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc [hunk 1]
```diff
@@ -37,13 +37,15 @@ Status SoftmaxOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
...
   if (input_size != 2) {
-    int32_t new_shape_0 = SafeInt<int32_t>(input_shape.data()[0]);
-    for (size_t i = 1; i < input_size - 1; i++) {
-      new_shape_0 *= input_shape.data()[i];
-    }
-    emscripten::val new_shape = emscripten::val::array();
-    new_shape.call<void>("push", new_shape_0);
-    new_shape.call<void>("push", static_cast<int32_t>(input_shape.back()));
+    NodeAttrHelper helper(node);
+    int32_t axis = helper.Get("axis", 1);
```
Comment (fdwr) [inline onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: 1 was the default axis for opset 11 https://github.com/onnx/onnx/blob/main/docs/Changelog.md#Softmax-11, but -1 is the default for opset 13+, but I'm not seeing any conditional opset selection for the default axis. I suppose it *technically* doesn't matter currently because `IsOpSupportedImpl` will return false, but I worry about leaving this here without comment or note, that it will be missed later.

`int32_t axis = helper.Get("axis", 1);  // Use the opset 13 default.`

Opset 13: https://github.com/onnx/onnx/blob/main/docs/Operators.md#Softmax
```
axis : int (default is -1)
Describes the dimension Softmax will be performed on. Negative value means counting dimensions from the back. Accepted range is [-r, r-1] where r = rank(input).
```

(Also we should support opset 13 eventually, fine to be a separate CR, even if we have to decompose it to separate WebNN calls : `Exp(input) / ReduceSum(Exp(input), axis=axis, keepdims=1)`)

Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: DirectML's [`DML_ACTIVATION_SOFTMAX1_OPERATOR_DESC`](https://learn.microsoft.com/en-us/windows/ai/directml/api/ns-directml-dml_activation_softmax1_operator_desc) supports opset 13, I am wondering if WebNN spec could support it as well? (One limitation is XNNPack only allows 2D input).

Comment (fdwr) [reply onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: > I am wondering if WebNN spec could support it as well?

Yeah, I think [WebNN `softmax`](https://www.w3.org/TR/webnn/#api-mlgraphbuilder-softmax-method) should, because [PyTorch `torch.nn.Softmax`](https://pytorch.org/docs/stable/generated/torch.nn.Softmax.html) and [ONNX `Softmax-13`](https://onnx.ai/onnx/operators/onnx__Softmax.html) and [TensorFlow `tf.nn.softmax`](https://www.tensorflow.org/api_docs/python/tf/nn/softmax) all accept an `axis` parameter, and WebNN's current operator cannot implement any of these. 😟

> DirectML's [DML_ACTIVATION_SOFTMAX1_OPERATOR_DESC](https://learn.microsoft.com/en-us/windows/ai/directml/api/ns-directml-dml_activation_softmax1_operator_desc) supports opset 13

Unfortunately the newer [DML_ACTIVATION_SOFTMAX1_OPERATOR_DESC](https://learn.microsoft.com/en-us/windows/ai/directml/api/ns-directml-dml_activation_softmax1_operator_desc) was added fairly recently in `DML_FEATURE_LEVEL_5_1`, which is beyond the Windows 11 target, and not even present in the version of the Windows SDK which Chromium uses. So we need to use the older operator, but the newer one can be implemented by the older one with the right combination of strides and/or transposes. Even multiple axes (the more generic form which is a superset of PyTorch/ONNX/TensorFlow's single axis form) could be implemented, but I'm not sure if any *other* ML API accepts multiple axes to necessitate that.

If XNNPack can be augmented to at least 3D (including a batch dimension, reduction dimension, and spatial dimension), then we could satisfy all those libraries with a trivial reshape, since any leading batch dimensions or trailing spatial dimensions can be folded together.

Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: Thanks @fdwr, that's informative, how about report this issue to spec then we support the opset 13 in WebNN EP by decomposing it to separate WebNN calls firstly?

Comment (fdwr) [reply onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: Added todo to create issue.

Decomposition in the meantime should work, but using transpose with reshape would likely be faster than multiple elementwise operations, to rearrange any middle dimension axes toward the end, and then to collapse/fold any leading dimensions via reshape.

Comment (Honry) [reply onnxruntime/core/providers/webnn/builders/impl/softmax_op_builder.cc:41]: Yeah, indeed, `transpose` + `reshape` should be faster.
